Chebato/setup phase

CheBato opened the pull request:
- Added the setup phase test suite

AMMayberry1 commented on test/helpers/IntegrationHelper.js:
```diff
@@ -836,22 +840,24 @@ global.integration = function (definitions) {
-                this.player1.setDiscard(options.player1.discard, ['removed from game']);
-                this.player2.setDiscard(options.player2.discard, ['removed from game']);
 
+                if (options.phase !== 'setup') {
```
This is a good start. One thing we need to consider though is that people might get confused and still try to set some of the configuration settings not realizing that they don't apply in setup phase tests. So it would be good to have some feedback mechanism for letting them know. 

Probably ought to be some kind of parameter checks where we throw an error if they've done something that doesn't apply in this type of test. This would apply not only for putting in cards in arenas like in this if block, but also if they try to set leader status deployed or base damage in the below settings.

I've honestly been meaning for a while to add some validation checks on the `options.playerX` config objects, mostly to catch if someone has a typo in their config and accidentally puts `groundAerna` or something because that would just silently fail. Maybe this is a good chance to add a validation method that can handle the normal checks for action phase tests and then the additional ones for setup phase tests as well.

CheBato replied:
hell yeah i'll write some up and you can check if they're okay cool?

AMMayberry1 replied:
Yup sounds good!

AMMayberry1 commented on test/server/core/gameSteps/phases/SetupPhase.spec.ts:
```diff
@@ -0,0 +1,100 @@
+                context.player1.clickPrompt('yes');
+                context.player2.clickPrompt('no');
+                const afterMulliganHand = context.player1.hand;
+                expect(beforeMulliganHand).not.toEqual(afterMulliganHand);
```
Also need to check that player2's hand has not changed

CheBato replied:
gotcha!

AMMayberry1 commented on test/server/core/gameSteps/phases/SetupPhase.spec.ts:
```diff
@@ -0,0 +1,100 @@
+                context.player2.clickCardNonChecking(context.player1.hand[0]);
+                context.player2.clickCardNonChecking(context.player1.deck[0]);
+                context.player2.clickCardNonChecking(context.p1Base);
+                context.player2.clickCardNonChecking(context.p2Base);
```
Need to add some expectation after these clicks to confirm they didn't have any effect. Probably something checking the prompt selection state?

CheBato replied:
yeah probably we can do that the prompt hasn't changed after each for both players.

AMMayberry1 commented on test/server/core/gameSteps/phases/SetupPhase.spec.ts:
```diff
@@ -0,0 +1,100 @@
+
+                // Select 2 correct cards to resource
+                context.player2.clickCard(context.player2.hand[0]);
+                context.player2.clickCard(context.player2.hand[1]);
```
So this is something I forgot - currently the setup phase resource prompt doesn't have a "Done" button, it just automatically resources once both players have clicked 2 cards. I think that's probably not what we want, it would be better to ask players to click "done" so they can lock in their choice.

Can you make that update to the ResourcePrompt? You can copy some of the logic from VariableResourcePrompt, or maybe even move it completely since ResourcePrompt is the parent class and VariableResourcePrompt will just inherit it.

Note that making this change will also require updating the test setup stuff since now the test setup will need to additionally click "Done" for each player

CheBato replied:
sure i can add the Done to the resource phase.

danbastin replied:
Agreed here. In general our approach is that multiple selection will require a confirmation while single selection automatically submits. Might revise that as we see more cases but certainly for this one we want a done button. Do we also have the ability to "un-resource" a card?

AMMayberry1 replied:
The resourcing doesn't happen until the prompt resolves so shouldn't be needed, if I'm understanding the intent right

CheBato replied:
if i understand @danbastin correctly its whether we support unselecting cards. If we don't, I can make an addition to the ResourcePrompts clickCard function to check whether the card is already in the selectedCards and remove it from there, if that makes sense.

AMMayberry1 commented on test/server/core/gameSteps/phases/SetupPhase.spec.ts:
```diff
@@ -0,0 +1,100 @@
+                expect(context.player1.resources.length).toBe(2);
+                expect(context.player2.resources.length).toBe(2);
+
+                // Check if resources are correctly set
```
Can we check for both players

CheBato replied:
ofcourse

AMMayberry1 commented on test/server/core/gameSteps/phases/SetupPhase.spec.ts:
```diff
@@ -0,0 +1,100 @@
+
+                // The setup phase is divided into 3 or 4 steps: choose Initiative player, Draw card step, mulligan step (optional), resource step
+                // Choose Initiative Step
+                context.selectInitiativePlayer(context.player1);
```
Can we add a second (smaller) test to confirm that it works if player 2 is the initiative player

CheBato replied:
sure no problem :)

AMMayberry1 commented on server/game/core/gameSteps/prompts/ResourcePrompt.ts:
```diff
@@ -8,18 +8,19 @@ import { AllPlayerPrompt } from './AllPlayerPrompt';
     ) {
         super(game);
         game.getPlayers().forEach((player) => this.selectedCards[player.name] = []);
+        game.getPlayers().forEach((player) => this.playersDone[player.name] = false);
```
Let's just roll these two into one regular `for` loop

AMMayberry1 commented on server/game/core/gameSteps/prompts/ResourcePrompt.ts:
```diff
@@ -65,10 +66,10 @@ export class ResourcePrompt extends AllPlayerPrompt {
 
-        if (!this.activeCondition(player)) {
+        if (!this.activeCondition(player) || !player.hand.includes(card) ||
+          this.selectedCards[player.name].length === this.nCardsToResource) {
```
Just for style consistency could you break out the lines like this:

```typescript
  if (
    !this.activeCondition(player) || !player.hand.includes(card) ||
     this.selectedCards[player.name].length === this.nCardsToResource
) {
```

CheBato replied:
ofcourse!

AMMayberry1 commented on server/game/core/gameSteps/prompts/ResourcePrompt.ts:
```diff
@@ -86,7 +87,20 @@ export class ResourcePrompt extends AllPlayerPrompt {
 
     public override menuCommand(player, arg): boolean {
-        Contract.fail(`Unexpected menu command: '${arg}'`);
+        if (arg === 'done') {
```
Can we still have the Contract.fail if the command is something else, since that would mean they somehow clicked a button that doesn't exist

CheBato replied:
Yessir!

AMMayberry1 commented on server/game/core/gameSteps/prompts/ResourcePrompt.ts:
```diff
@@ -86,7 +87,20 @@ export class ResourcePrompt extends AllPlayerPrompt {
+                return false;
+            }
+
+            this.resourceSelectedCards(player);
```
If I'm reading this right, looks like it will resource the cards for a player as soon as they click done regardless of the other player's finished state. I think we should wait to resource until both players are done.

CheBato replied:
sure i'll change it to when they both complete it!

AMMayberry1 commented on test/helpers/GameFlowWrapper.js:
```diff
@@ -56,6 +56,7 @@ class GameFlowWrapper {
     resourceAnyTwo() {
         this.guardCurrentPhase('setup');
         this.allPlayersInInitiativeOrder().forEach((player) => player.clickAnyOfSelectableCards(2));
+        this.allPlayersInInitiativeOrder().forEach((player) => player.clickPrompt('Done'));
```
Similar comment as above - let's just roll these two into one `for` loop

CheBato replied:
yessir!

AMMayberry1 commented on test/helpers/IntegrationHelper.js:
```diff
@@ -754,6 +754,42 @@ function generatePromptHelpMessage(player) {
 
+function validatePlayerOptions(playerOptions, playerName, startPhase) {
+    // list of approved property names
+    const listOfApprovedPropertiesForAnyOtherPhase = [
```
Let's say "ForNonSetupPhase" so this is understandable without seeing the other variable name. Also, it's personal preference but I think "listOf" in the name is kind of uncessary, if the name is plural it implies a list imo

CheBato replied:
ah yes thank you i was thinking how i can rename this sausage of a variable.

AMMayberry1 commented on test/helpers/IntegrationHelper.js:
```diff
@@ -754,6 +754,42 @@ function generatePromptHelpMessage(player) {
+
+    // Check for unknown properties
+    if (startPhase !== 'setup') {
+        Object.keys(playerOptions).forEach((prop) => {
```
Per the discussion in discord, for anything that's more than a one-liner I would prefer to use `for` loops since it's better for debugging (the vscode debugger skips directly over forEach loops). For one-line statements I'm more okay with forEach but there's still some discussion on it

CheBato replied:
i also made it a bit more clearer and cleaner.

AMMayberry1 commented on test/helpers/IntegrationHelper.js:
```diff
@@ -799,6 +835,11 @@ global.integration = function (definitions) {
+
+                // validate supplied parameters
+                validatePlayerOptions(options.player1, 'player1', options.phase);
+                validatePlayerOptions(options.player2, 'player2', options.phase);
```
Awesome!! Thanks so much, been wanting this for a while

CheBato replied:
With pleasure!

AMMayberry1 commented on test/helpers/IntegrationHelper.js:
```diff
@@ -826,35 +867,46 @@ global.integration = function (definitions) {
-                this.player1.damageToBase = options.player1.damageToBase ?? 0;
-                this.player2.damageToBase = options.player2.damageToBase ?? 0;
+                    // Player stats
+                    this.player1.damageToBase = options.player1.damageToBase ?? 0;
```
Could we handle this inside the base setup method instead of needing to manage it explicitly here? I feel like that'd be cleaner and easier to maintain

CheBato replied:
yessir!

CheBato replied:
Now that I have tried this quite a few times, it seems that this.player1.damageToBase needs to be set here; otherwise, some tests fail to end the game. I assume it has something to do with which player1 or player2 variable we are setting here. If you want you can take a look at it.

AMMayberry1 commented on test/server/core/gameSteps/phases/SetupPhase.spec.ts:
```diff
@@ -0,0 +1,185 @@
+
+                // Check that player1 cannot select any additional cards
+                context.player1.clickCardNonChecking(context.player1.hand[2]);
+                expect(context.player1.selectedCards.length).toBe(2);
```
These tests are awesome!! Do you mind also adding similar checks of the card selection process to the resource step of the setup phase, to the degree it makes sense?

CheBato replied:
do you mean the setupPhase or the regroup phase?

CheBato replied:
Added some additional ones in the setupPhase for good measure and some for the regroup Phase.

danbastin commented:
In the `resourceSelectedCards` the `resourceCard` method should be `player.resourceCard(card, false);` to ensure that the cards are ready when resource and available on the first turn. The test should check for this as well.

CheBato commented:
> In the ResourcePrompt the `resourceSelectedCards` the `resourceCard` method should be `player.resourceCard(card, false);` to ensure that the cards are ready when resource and available on the first turn. The test should check for this as well.

I changed the player.resourceCard(card) -> player.resourceCard(card, false); and added tests to verify they are indeed unexhausted.

AMMayberry1 commented on test/helpers/IntegrationHelper.js:
```diff
@@ -769,24 +769,18 @@ function validatePlayerOptions(playerOptions, playerName, startPhase) {
     ];
     // list of approved property names for setup phase
-    const listOfApprovedPropertiesForSetupPhase = [
+    const ForSetupPhase = [
```
nit: should start with lowercase
